server/container/util-java8/src/main/java/org/apache/james/util/FluentFutureStream.java

```diff
@@ -69,11 +80,28 @@ private FluentFutureStream(CompletableFuture<Stream<T>> completableFuture) {
             CompletableFutureUtil.thenComposeOnAll(completableFuture(), function));
     }
 
+    public <U> FluentFutureStream<U> thenFlatCompose(Function<T, CompletableFuture<Stream<U>>> function) {
```

Comment:
Ok let's merge this PR in this state and then introduce Scala into James: http://stuq.nl/weblog/2008-11-26/4-steps-to-add-scala-to-your-maven-java-projects. It look like we can do that progressively.
I think all the team will be excited about that and we will have real monad support!